### ISSUE TITLE
[Feature Request] Recall Matrix like in SuperMemo algorithms

### ISSUE THREAD

**Issue by Expertium** [open-spaced-repetition/fsrs4anki#271]
**Which module is related to your feature request?**
Scheduler, Optimizer

**Is your feature request related to a problem? Please describe.**
It's a way to correct out theroretical function using real data to obtain more accurate predictions in cases where the theoretical function makes poor predictions.

**Describe the solution you'd like**
[Recall Matrix.docx](https://github.com/open-spaced-repetition/fsrs4anki/files/11898414/Recall.Matrix.docx)
It's a pretty long description, but that's because it's a very sophisticated feature. However, it oculd potentially have a greater impact on the accuracy than any other change so far.

**Additional context**
While I tried to make the description as detailed and clear as possible, it's possible that the final implementation will be different. Still, this is about as good of a description as I can make. It's fine if this isn't your top priority right now, but I highly recommend implementing it in the foreseeable future.

**Comment by L-M-Sherlock** [comment]
Let's discuss this issue here.

> The recall matrix should be a 3-dimensional matrix, where the dimensions are S, D, and theoretical R. Similar to the B-W heatmap, but with theoretical R as a new dimension.

Why use the theoretical R as an extra dimension? Why not just save the measured R in the matrix entry like this: `R_M = RecallMatrix(D_rounded, S_rounded)`

**Comment by Expertium** [comment]
Because the same S could correspond to different R because of t. If, for 2 predictions, S is the same but t is different, R will be different. Grouping them together will reduce accuracy.

**Comment by L-M-Sherlock** [comment]
<img width="1054" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/aee2f861-66af-49c0-98a0-9e8f30d4061b">

```python
B_W_Metric_raw['r_bin'] = B_W_Metric_raw['p'].map(lambda x: round(x * 4, 1) / 4)
R_Matrix = B_W_Metric_raw.groupby(by=['s_bin', 'd_bin', 'r_bin']).agg({'y': ['mean', 'count']}).reset_index()
R_Matrix.columns = ['S_rounded', 'D_rounded', 'R_t_rounded', 'R_measured', 'count']
R_Matrix[R_Matrix['count'] > 3]
```
Paste it after the B-W-Metric cell.

**Comment by L-M-Sherlock** [comment]
```python
DiffMax = max(abs(R_Matrix['R_measured']-R_Matrix['R_t_rounded']))
```

**Comment by Expertium** [comment]
It doesn't show anything, other than an indicator that the code has been executed successfully.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/19f429d7-ba2f-483a-b881-8abb77266719)

**Comment by L-M-Sherlock** [comment]
You should print() them. I have updated the mini-batch.ipynb: https://colab.research.google.com/github/open-spaced-repetition/fsrs4anki/blob/Expt/new-baseline/experiment/mini-batch.ipynb

**Comment by Expertium** [comment]
Yes, looks good.

**Comment by Expertium** [comment]
I know it's too early to think about it, but I thought about whether we should
1) Allow users to have different sets of parameters for different decks and different recall matrices for each set
2) Allow users to have different sets of parameters for different decks but use the same recall matrix
3) Allow users to only have 1 set of parameters and 1 recall matrix, and remove the ability to specify different sets of parameters for different decks

I think 2 is the best option. If we assign a different recall matrix to each different set of parameters, those matrices will be very sparse. For example, suppose that the entire collection has 100 000 reviews, but the user decides to add a new set of parameters for a deck that only has 3 000 reviews. The recall matrix for the entire collection will have a lot of information and a lot of entries, whereas the recall matrix for that one deck will have very few entries with low counts, making it less useful. The more data, the better. There is no upside to using a sparse recall matrix based on little info.
At the same time, not allowing users to have different parameters for different decks would reduce the flexibility of FSRS. So yeah, 2 seems like the best option.

**Comment by Expertium** [comment]
@L-M-Sherlock seems like it's finally time to start working on this? I believe all other important issues have been resolved.

**Comment by L-M-Sherlock** [comment]
There are still some bugs in the Helper add-on. I need to deal with them firstly.

**Comment by L-M-Sherlock** [comment]
Sorry for the late reply. I'm busy for attending an academic conference in few days later. And the feature requests is overflowing. The bug fixing is in high priority. I will focus on the development of R-metric after July 3.

**Comment by L-M-Sherlock** [comment]
I have a question: how to optimize $w_1$ and $w_2$?

**Comment by Expertium** [comment]
Oh yeah, about that. I believe we should completely change our loss function. I've seen a problem multiple times - changing formulas decreased log-loss by around 1%, but RMSE decreased by 5%/10%/20%.
I think we should minimize the difference between theoretical R and R-matrix entry. In other words, the new loss will be WozLoss(Rt, Rm), or ln(1-(abs(Rt-Rm)))

**Comment by Expertium** [comment]
Wait, I just realized - we can't optimize w1 and w2 that way, since those are used in weighted R.
Ok, forget what I said above then. Let's not change the loss function for now. Just make the algorithm output Rw instead of Rt whenever it predicts R.
By the way, once the R-matrix is done, every single calculation involving R (in the optimizer, scheduler and helper) will have to be based on the weighted R. That will involve changing a lot of the code, especially for the Helper add-on.

**Comment by L-M-Sherlock** [comment]
Before we use the R-Matrix, we should train the parameters of FSRS at first. How can we optimize w1 and w2 in the same way as other parameters in the same time?

**Comment by Expertium** [comment]
I don't see why we can't optimize them simultaneously. For each epoch we make the R-matrix, calculate weighted R and use it in the loss function, clear the R-matrix and repeat this for the next epoch.

**Comment by L-M-Sherlock** [comment]
> I don't see why we can't optimize them simultaneously. For each epoch we make the R-matrix, calculate weighted R and use it in the loss function, clear the R-matrix and repeat this for the next epoch.

Which parameters should we use to generate R-matrix? The parameters at the start of the epoch?

**Comment by Expertium** [comment]
That's a good question, I haven't thought about that. I think using parameters from the previous epoch (init_w if this is the first epoch) is reasonable.

**Comment by user1823** [comment]
@Expertium, what if the weight comes out to be very close to 1?

I think that the optimizer would try to do this because when weight is equal to 1, the loss becomes very small.

**Comment by Expertium** [comment]
W1 determines the maximum weight, and we will let the optimizer choose it. We'll see if that leads to such issues, for now it's hard to predict how it will behave. And even if sets w1 to 0.99 (that's how I plan to clamp w1), I don't think this is bad, since it will be 0.99 only for entries based on a lot of reviews, in other words, it will be 0.99 for R that is estimated very accurately, which is exactly what we want.

**Comment by user1823** [comment]
I thought about it more thoroughly and realized that the model would still need to accurately classify the cards according to their stability and difficulty in order to reap good benefits of the R-matrix. So, there shouldn't be any major problem. But, let's see how it goes.

**Comment by L-M-Sherlock** [comment]
https://colab.research.google.com/github/open-spaced-repetition/fsrs4anki/blob/Expt/new-baseline/experiment/mini-batch_R-Matrix.ipynb

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/f79f34e7-4efd-481c-a633-e6d697194963)

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/fb7a2f85-ff76-4cc1-ab9c-2e6a3ef3be29)

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/3b07d436-7c57-4dc6-92fe-eae356cf99cf)

**Comment by L-M-Sherlock** [comment]
> 4\. While you have implemented R_w, you haven't used it to its full potential. As described in my .docx document, Rw should be used to calculate a new estimate of the **current** S, and also in the formula for new_s.

The new estimate of the current S is not differentiable because we calculate it via indexing the R-Matrix. The gradient descent method is blocked by it.

**Comment by L-M-Sherlock** [comment]
> 1. I would advise you to add this code to 4.0 alpha, since that's what I'm using as the baseline, and I want to have a fair comparison (same goes for the hybrid S0 curve-fit/pre-train method in the other issue).

The weight to average R-Matrix has achieved 0.99, which means the model structure of FSRS is unimportant. The output of FSRS is ignored. So it would output the same result whatever I implement it in any other baselines.

**Comment by L-M-Sherlock** [comment]
> 3. We might need a few more epochs for it to converge, since the R-matrix is always kinda "lagging behind" due to being calculated using the parameters of the previous epoch.

The current version is using the parameters in the end of the epoch. Because I find that it is impossible to optimizer the parameters of FSRS and the weights for R-matrix in the same time. And when I used the parameters from the previous epoch, the indexError occurred frequently.

**Comment by Expertium** [comment]
> The weight to average R-Matrix has achieved 0.99, which means the model structure of FSRS is unimportant. The output of FSRS is ignored. So it would output the same result whatever I implement it in any other baselines.

Not really, since there is another parameter and the weight depends on the number of entries and the second parameter.

**Comment by Expertium** [comment]
> The new estimate of the current S is not differentiable because we calculate it via indexing the R-Matrix.

What about just passing R_w to `stability_after_success` and `stability_after_failure` then?

**Comment by L-M-Sherlock** [comment]
> What about just passing R_w to `stability_after_success` and `stability_after_failure` then?

It will break the gradients of chain of stabilities in [Backpropagation through time](https://en.wikipedia.org/wiki/Backpropagation_through_time). In original FSRS, the loss will back propagate in the opposite direction in the below figure:

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/99bd16b6-bc95-414b-94a1-c46e3d756c33)

If we replace r with R_w, a path of backpropagation will be disable:

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/5a38b9c6-094a-40f2-add9-8818f825d812)

**Comment by Expertium** [comment]
> > The weight to average R-Matrix has achieved 0.99, which means the model structure of FSRS is unimportant. The output of FSRS is ignored. So it would output the same result whatever I implement it in any other baselines.
> 
> Not really, since there is another parameter and the weight depends on the count and the second parameter.

I'm currently running the optimizer, I'll see what parameters I get. If the second parameter (that power that the ratio of counts is raised to) won't reach it's minimum or maximum value, then it means that we just simply need to remove the other parameter as useless and only keep the power parameter.

**Comment by Expertium** [comment]
Here are the optimal parameters for me. I guess we simply don't need the first parameter for the weight.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/a51f30e1-caa9-44bb-9519-672bbf7d45a5)
By the way, this is what it looks like if you plot it. Here x is the raw ratio min(n, 300)/300
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/5846436c-f6fc-4b3d-8fe1-0bbe47b43cdb)

And I keep running out of RAM, not just in this version (LSTM and some other version also had this problem):
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/ea52d44e-4365-4378-992a-197b14d144db)

**Comment by Expertium** [comment]
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/86265735-542b-4d02-bd5b-82969453d337)

**Comment by Expertium** [comment]
I set revlog_start_date = "2022-07-04" to get rid of the RAM issue
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/81d598e7-1968-4020-b81f-5193671a432d)

**Comment by Expertium** [comment]
> If the optimizer determined that the best course of action is to always assign the maximum possible weight to the R-matrix entry, it would set w[13] to 0.01, effectively making it close to 1 for any count (n). But that's not what's happening.
> In your collection it hasn't reached its minimum value.

And in my collection it's around 0.2, so yes, the optimizer doesn't think that assigning the maximum possible weight to the R-matrix entry in every single case is optimal.
But if we can't incorporate R_w into formulas for S in any way due to problems with backpropagation, then this entire idea is useless anyway, since instead of correcting S we are just masking incorrect predictions of S and sweeping them under the rug.

**Comment by user1823** [comment]
For me, the optimizer with the R-matrix produced the best results till now. The RMSE is just 0.0068!!

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/c69d0f4f-ff07-4c88-88e2-bc3df6a81b28)

My parameters:
[1.5942, 2.4084, 4.3713, 1.956, 1.9538, 0.0003, 1.8556, 0.1842, 1.2342, 1.764, 0.506, 0.8366, 0.1326, 0.872, 0.5851]

So, w[13] is not always 0.99.

> But if we can't incorporate R_w into formulas for S due to problems with backpropagation, then this entire idea is useless anyway, since instead of correcting poorly predicted S we are just sweeping the problem under the rug.

We can't do it in the optimizer. But, in my opinion, we can still do it in the scheduler. Now, because the optimizer says that the RMSE is reduced by a large amount when R is corrected using the R-matrix, I think that it will work well if we incorporate it into the scheduler.

> The weight to average R-Matrix has achieved 0.99, which means the model structure of FSRS is unimportant. The output of FSRS is ignored. 

I also thought this previously. But, this is not the case even if the weight comes out to be very close to 1. FSRS model is still being used to classify reviews based on their stability and difficulty. If we change the structure of the model, the reviews would be classified differently. This would mean that the values in the R-matrix are different. So, the structure of the model still affects the calculation of the retention.

**Comment by Expertium** [comment]
@L-M-Sherlock I did more tests, and for 7 of my decks w[12] was below 0.9. Btw, I use 5 epochs and 5 splits. Also, w[13] never reached 0.01 for any deck, once again suggesting that weight based on the count was a good idea and the optimizer won't just assign maximum weight to every R-matrix entry.

[0.3934, 1.0303, 4.6616, 2.4349, 1.8984, 0.08, 1.672, 0.2027, 1.0563, 1.7025, 0.5535, 0.7508, 1.0757, **0.898**, 0.444]
[0.2891, 0.409, 5.158, 1.3497, 1.2838, 0.0773, 1.4818, 0.3212, 0.9294, 1.7045, 0.5219, 0.7339, 0.9873, **0.834**, 0.6425]
[0.275, 0.3638, 5.3607, 1.1494, 1.2129, 0.0722, 1.1637, 0.1556, 0.6534, 1.6455, 0.5417, 0.623, 0.9995, **0.8364**, 0.6207]
[1.9826, 2.2212, 4.6363, 1.414, 1.179, 0.05, 1.9911, 0.15, 1.4366, 1.8329, 0.3743, 0.3405, 0.599, **0.8433**, 0.6313]
[0.2149, 0.9368, 5.052, 2.2627, 1.6468, 0.0574, 1.7516, 0.2077, 1.1594, 1.7728, 0.4884, 0.8043, 1.2465, **0.627**, 0.8053]
[0.2847, 0.3138, 5.3654, 1.3049, 1.0671, 0.0734, 1.2429, 0.1534, 0.6912, 1.6494, 0.5355, 0.6251, 0.7018, **0.8172**, 0.6647]
[0.2135, 0.2479, 5.5107, 1.4421, 1.1158, 0.0729, 1.2866, 0.2053, 0.7504, 1.6562, 0.5287, 0.7106, 1.1989, **0.834**, 0.6067]

And then I noticed that w[14] isn't clamped. I clamped it...and it completely changed all behaviour. Now w[14] is always 0.01. So everything I posted above (especially the sets of parameters in this comment) is invalid, because somehow clamping changed what is optimal. No idea how. And I also got confused about which parameter is which, so when I said w[13] I meant w[14] and when I said...ok, you know what, my brain is fried.

**Comment by Expertium** [comment]
> We can't do it in the optimizer. But, in my opinion, we can still do it in the scheduler. Now, because the optimizer says that the RMSE is reduced by a large amount when R is corrected using the R-matrix, I think that it will work well if we incorporate it into the scheduler.

We won't be able to asssess exactly how accurate such a scheduler will be, since it's impossible to run it in the optimizer to get RMSE.

**Comment by user1823** [comment]
> > We can't do it in the optimizer. But, in my opinion, we can still do it in the scheduler. Now, because the optimizer says that the RMSE is reduced by a large amount when R is corrected using the R-matrix, I think that it will work well if we incorporate it into the scheduler.
> 
> We won't be able to assess exactly how accurate such a scheduler will be, since it's impossible to run it in the optimizer to get RMSE.

Well, we can. We just need to analyze the reviews using the final parameters and the final R-matrix after the training is complete.

**Comment by Expertium** [comment]
You mean we optimize one verison of the algorithm (one where R_w isn't used to calculate S), and then we run a different version (one where R_w is used to calculate S), and then calculate RMSE using the second version? That's a really strange way of doing it, butit might just somehow work.

**Comment by L-M-Sherlock** [comment]
I think R-Matrix has good performance but it doesn't has generality. For example, I use this code to re-calculate the stability:

<img width="848" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/58104844-857e-4cef-965f-3dd710626301">

Some weird things could happen: when p decays with delta_t, the r_w might not decay monodically. If we recalculate the stability based on r_w, the stability would vary over time.

**Comment by Expertium** [comment]
We know it's _somehow_ possible to make it work, since SM-17 uses it. The question is how exactly.

**Comment by L-M-Sherlock** [comment]
Here is an example in my collection:

<img width="694" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/86d09bc0-735c-4345-a0ae-a314e6150292">

**Comment by Expertium** [comment]
I would say that if accuracy (RMSE and the universal metric) are significantly improved, then we should accept non-monotonicity as a quirk of the algorithm. I can't really think of any way to solve.
Currently lack of backpropagation for the R-matrix is the largest obstacle. Then there is the weirdness with clamping, where clamping w[14] completely changes the final parameters. And then there is lack of monotonicity for S_w. If we don't solve the first problem, lack of backpropagation, then we might as well scrap this idea.

**Comment by L-M-Sherlock** [comment]
> Currently lack of backpropagation for the R-matrix is the largest obstacle. Then there is the weirdness with clamping, where clamping w[14] completely changes the final parameters. And then there is lack of monotonicity for S_w. If we don't solve the first problem, lack of backpropagation, then we might as well scrap this idea.

To optimize it with the gradient descent method, it should have [subgradient](https://en.wikipedia.org/wiki/Subgradient_method) at least. But R-Matrix[D,S,R] is even not a continuous function. It doesn't have gradient or subgradient. I can't figure out anyway to solve the obstacle.

**Comment by Expertium** [comment]
There are papers detailing other approaches (alternatives to backpropagation), but I don't think pytorch supports anything else. So unless you want to code your own optimizer after reading scientific papers, you can close this issue.

**Comment by L-M-Sherlock** [comment]
Maybe Woz know that, but he hasn't published any papers about his algorithm for a long time. What do you think of the idea bout R-Matrix? @user1823

**Comment by Expertium** [comment]
He suggested optimizing a version of the algorithm where S is not based on R_w, but then using a version where S is based on R_w in the scheduler. And RMSE would be calculated usign the second version. The problem is that there is now way to ensure that parameters that are optimal for the first version are also optimal for the second version.

**Comment by Expertium** [comment]
> Maybe Woz know that, but he hasn't published any papers about his algorithm for a long time.

I don't think he even published any papers at all. So here is all the info we have if we want ot reverse-engineer the R-matrix
https://supermemo.guru/wiki/Recall_matrix
https://supermemo.guru/wiki/Algorithm_SM-17#Retrievability

**Comment by Expertium** [comment]
If nothing works, we can implement grade-derived R as a "consolation prize". I figured out a good way to obtain R from grades using curve_fit, and those values can be precomputed before the optimization. Obviously even in the absolute best case scenario it won't be as good as the R-matrix since it provides, as most, 4 unique values of R, but that's why I'm calling it a consolation prize.

**Comment by Expertium** [comment]
@L-M-Sherlock @user1823 I tested everything `again`, with clamped R_w and clamped w[14]. I'm also using 5 splits and 5 epochs.
```
w[14] = w[14].clamp(0.01, 2)
r_w = r_w.clamp(0.0001, 0.9999)
```
The results are very strange.
It works in most cases, but not always,
Here are some of sets of parameters for my decks and collections of other users.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/0d06e807-a6d7-4b4b-a9b2-c279262b1110)
Optimization "failed" for Sherlock and user1823: both parameters reached extreme values. Removing the clamping for w[14] while keeping the clamping for R_w makes it possible for this parameter to become negative, which we obviously don't want.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/33882998-1913-47d4-abf8-3fd4b46dee09)
Removing the clamping for R_w and keeping it for w[14] doesn't help. I was unable to reproduce [these results](https://github.com/open-spaced-repetition/fsrs4anki/issues/271#issuecomment-1620390896). If the clamping/number of epochs/number of splits completely changes the final parameters, then it's better to not use this at all. It means that the parameters are not robust and extremely sensitive to small changes in the data and/or optimization procedure. I think we should close this issue as a failure.

**Comment by L-M-Sherlock** [comment]
<img width="669" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/c534248f-0e93-4339-b352-c8af3b508d7e">

By the way, I find that even with R-matrix, the green line still depart from the orange line in your calibration.

**Comment by nb9618** [comment]
> If the clamping/number of epochs/number of splits completely changes the final parameters, then it's better to not use this at all. It means that the parameters are not robust and extremely sensitive to small changes in the data and/or optimization procedure.

@Expertium I should say that for my deck the result of optimization of the baseline optimizer varies a lot with number of splits and epochs, so this problem is probably general to the optimizer, not only to this variation.

**Comment by Expertium** [comment]
@L-M-Sherlock I have a question - if w_1 and w_2 are fixed (not optimizable), would it then be possible to use the R-matrix to calculate weighted R and use it to recalculate S and calculate new S, or no?

**Comment by L-M-Sherlock** [comment]
> @L-M-Sherlock I have a question - if w_1 and w_2 are fixed (not optimizable), would it then be possible to use the R-matrix to calculate weighted R and use it to recalculate S and calculate new S, or no?

I don't know. I encourage you to do these research.

**Comment by Expertium** [comment]
That would be very difficult for me to code. I would greatly appreciate if you helped me.
Basically, implement it as I described in my .docx document, but set w_1=0.9 and w_2=0.5.

**Comment by Expertium** [comment]
I tested [this](https://colab.research.google.com/drive/1ROKhfPQ334ttz1vHpH_ABFWucWchBrDj?usp=sharing). I only made 2 changes to the code - changed clampings for w[14] and clamped the count (that latter is quite important).
It took almost 4 hours to optimize it on my entire collection, and then something weird happened and it got stuck.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/c38b3279-c960-46b1-8697-42bbeb5bfb70)
The parameters seemed reasonable though.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/141b3e9a-e77b-4a5e-85d1-4960515f1555)
And according to my very limited testing on 2 of my decks and user1823's collection, it is better than baseline. So this definitely has some potential, if only the optimization was faster.

**Comment by user1823** [comment]
> I don't know how user1823 got RMSE>0.05 for his collection.

I tested it before Sherlock made some important changes to the calculation of loss.

**Comment by L-M-Sherlock** [comment]
I update the code to set the upper limit for count.

**Comment by L-M-Sherlock** [comment]
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/5d3e0ed6-76c5-4ecf-aab4-4e2f69eb9efd)

The intervals are weird.

**Comment by Expertium** [comment]
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/ca46dd33-0d76-4297-a2dd-6f6211af717e)
Here are the results on my colelction, using https://colab.research.google.com/github/open-spaced-repetition/fsrs4anki/blob/Expt/new-baseline/candidate/R-Matrix.ipynb (the version I copied and edited myself produced different results for some reason)
The intervals are a little weird, I agree.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/36988cab-9e96-4c15-b184-3b20af1615f0)

**Comment by Expertium** [comment]
So, SHerlock, do you have any ideas how to speed up optimization? Also, I recommend re-opening this issue.

**Comment by user1823** [comment]
> The intervals are a little weird,

Do you think that this is a "little" wierd? The intervals not increasing beyond 1.7 months is a matter of significant concern.

> Also, I recommend re-opening this issue.

Reopening the issue is worthwhile only if we can think of a solution for these wierd intervals.

**Comment by Expertium** [comment]
> Do you think that this is a "little" wierd? The intervals not increasing beyond 1.7 months is a matter of significant concern.

Fair. I'm just more surprised by Sherlock's results, he gets the shortest max interval when "Good" is the first rating. That is really puzzling.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/81c97e8d-5898-457a-a7dc-3643d3a9ce20)

**Comment by L-M-Sherlock** [comment]
> So, Sherlock, do you have any ideas how to speed up optimization?

The main bottleneck is querying R-Matrix when FSRS predicts the memory states from the entire review history. If the sequence has 10 repetitions, it requires 10 queries. I have no idea how to speed it up.

**Comment by Expertium** [comment]
@L-M-Sherlock let's give this another try, but in a different way.
Instead of grouping reviews by D, S, and R, we will group them by delta_t (interval length), n (number of reviews), and average grade (just a simple arithmetic mean of grades). The advantage is that such a matrix can be precomputed from raw data, and we don't need any model of memory to compute it. It will be less accurate that way, but hopefully it will have fewer issues. The weighting system can remain the same.

**Comment by L-M-Sherlock** [comment]
> Instead of grouping reviews by D, S, and R, we will group them by delta_t (interval length), n (number of reviews), and average grade (just a simple arithmetic mean of grades).

How to index this matrix during training? `delta_t` and `n` existed for the single review. But where is `average grade`?

**Comment by Expertium** [comment]
We can just keep track of the average grade and pass it to state.

**Comment by Expertium** [comment]
I have an idea how to improve accuracy even further by introducing another matrix, S-matrix, but we need to test this new R-matrix first, to see if it works. At the very least it should be faster since it only needs to be filled once, unlike the previous version which had to be recalculated during training.

**Comment by L-M-Sherlock** [comment]
https://github.com/open-spaced-repetition/fsrs4anki/blob/Expt/r-matrix/archive/candidate/r-matrix-t-n-g.ipynb

I implement the simple version based on 4.0.0. Could you help me test it?

**Comment by Expertium** [comment]
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/151b6ae9-7c0e-48f7-a1d8-b61459537450)
I can't open it in collab

**Comment by L-M-Sherlock** [comment]
https://colab.research.google.com/github/open-spaced-repetition/fsrs4anki/blob/Expt/r-matrix/archive/candidate/r-matrix-t-n-g.ipynb

**Comment by L-M-Sherlock** [comment]
I'm updating the code. You can replace

```python
df['lapse'] = df['r_history'].str.count('1')
```

with 

```python
df['lapse'] = df['r_history'].astype(str).str.count('1')
```

to solve it.

**Comment by Expertium** [comment]
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/f299371e-baa6-4335-83ec-1f9dadf5e03f)
Now I'm getting the same error as in the beta version.

**Comment by L-M-Sherlock** [comment]
Maybe you can switch to another deck?

**Comment by Expertium** [comment]
I mean, yes, I can, but it's still a problem that has to be addressed. Ok, I will test it on other decks.

**Comment by Expertium** [comment]
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/6c60432b-f098-48e4-bf02-6898b863fdc0)
The optimization on my entire collection isn't done yet, but I can already tell that it behaves differently compared to the previous implementation: the parameter that determines the maximum amount of weight that can be allocated to the R-matrix entry is nowhere neat 0.99. It actually seems pretty fairly balanced.

**Comment by Expertium** [comment]
@user1823 you should try the new version too: https://github.com/open-spaced-repetition/fsrs4anki/issues/271#issuecomment-1634149318
I'm curious to see if you will get any kind of strange or unexpected results.

**Comment by Expertium** [comment]
For my collection RMSE went down only by about 13% compared to 4.0.0 beta. from 0.0589 to 0.0515; nowhere near as much as I was expecting.

**Comment by Expertium** [comment]
I think grouping should be improved.
1) Use the same (or slightly modified) formula that you used to round S to round delta_t. If one card has delta_t=365, and the other delta_t=366, it doesn't make much sense to put them into different categories.
2) Also group `i` differently. For example, i=20 and i=21 can safely be put into the same group.

**Comment by Expertium** [comment]
I tested this version, it performs better than v4 by around 9-10%, and it's statistically significant. I bet it can be improved with better grouping (see my comment above).

I could describe how to implement the S-matrix, my new idea, but considering that Sherlock already made up his mind about releasing v4 as is, I don't see the point of working on the S-matrix right now. I think we should do some final tests of this version of R-matrix, with better grouping, and leave it until Sherlock decides to start working on v5. Of course, if Sherlock decides to postpone the release of v4 to implement matrices, I would have nothing against it; quite the opposite, if I were in his shoes, I wouldn't release v4 before exhaustively trying everything related to matrices. But that would also require very major changes to the code of the helper add-on. And if it's true that FSRS will soon be supported on mobile, then this will increase the amount of headache tenfold or even make using matrices outright impossible, meaning that the mobile will need its own "FSRS Mobile" version with no matrices.

**Comment by L-M-Sherlock** [comment]
According my recent experiments, FSRS has reached the level of SM-15:

https://github.com/open-spaced-repetition/fsrs-vs-sm15/blob/main/evaluate.ipynb

And as we known, SM-15 employs many matrixes to fit the memory of user. FSRS v4 only uses 17 parameters for that. I think the benefit of matrixes couldn't compensate the cost to implement them.

If you no longer insist on implementing matrices in FSRS, I will close this issue.

**Comment by Expertium** [comment]
We still haven't properly implemented them though. For example, the last attempt had a major flaw with grouping reviews.

That being said, if FSRS is already more accurate than SM algorithms, we can put off the matrices and work on them in the future.

**Comment by Expertium** [comment]
While I would like to try a proper implementation of the R-matrix*, even if it performs well, I still don't know how to solve the problem with non-monotonicity of average R (predicted + R-matrix entry), and even if we somehow solve that, the matrix will be difficult to implement into Anki. So I suppose we can close this issue.

*with improved grouping. Last time cards with delta_t different by 1 day, like delta_t=365 and delta_t=365, were being grouped into distinct groups for no good reason.
